newIDE/app/src/MainFrame/ElectronMainMenu.js

```diff
@@ -67,82 +61,71 @@ type RootMenuTemplate =
  * Forward events received from Electron main process
  * to the underlying child React component.
  */
-class ElectronMainMenu extends React.Component<Props, {||}> {
-  _editor: ?MainFrame;
+class ElectronMainMenu extends React.Component<MainMenuProps, {||}> {
```

Comment:
Everything seems to be good, at least to me. Can you refactor `ElectronMainMenu` into functional component? It should be pretty straight forward at this point :) 
If not, I'd make a new PR. Your work has helped me a lot! Thanks :)

Reply:
Sure, that will be easy! I'll refactor it today.

Reply:
Sounds good 👍 
Once this is done and merged, this should help others PRs :)

Reply:
Okay, this turned out to be tougher than expected :sweat_smile:. I tried it and got it to work correctly, but the React warnings about missing hook dependencies are much harder to fix. Doing what the warning suggests leads to lots of unnecessary menu building calls.

I'm taking up a more involved approach now, so if it does not block any other PR, _please_ give me another day - I've always wanted to understand functional components properly and this task seems perfect for it. Else if it's blocking your PR @HarsimranVirk, you can take it up.

Really, _really_ sorry for the delay!

Reply:
It's alright, @NilayMajorwar, take your time :) I don't know if it helps or not, but try wrapping `_buildAndSendMenuTemplate` in a useCallback hook and pass `props` as a dependency. I don't think most of the props will change, and the ones which will change should trigger a re-render. This might help in resolving warnings (or not), either way there's nothing wrong in trying once :)

Reply:
There's a serious issue with this PR that I just discovered while trying to refactor - due to some reason, the ElectronMainMenu component gets unmounted and mounted each time the language or project changes. This causes the IPC renderer to set up repetitive listeners, which turns into a disaster whenever the user clicks on a menu item (one click on Open Project leads to 11 dialogs :). I'll fix this first, and then will continue refactoring.

Reply:
Thanks for your thorough check and finding this.
Is this when doing a functional component?

Because when props changes, components are *NOT* unmounted and remounted :) They are *re-rendered*. In case of:
* functional components, this means the function itself is launched. This means that you have to be careful of doing stuff like listeners only in useEffect. In particular, you may want to unregister listeners in the function that your return at the end of a useEffect:

  ```
  useEffect(() => { 
    // Register stuff
    return () => { /* Unregister stuff*/ }
  });
  ```
* for class components, props update means that `render` is called again, and then `componentDidUpdate`.

Reply:
Yeah, I know. But it _is_ happening for class components (confirmed, with a console.log in `componentWillUnmount`, and also all the `MaxListenersExceededWarning`s after few language changes). On some investigation, it seems like ElectronMainMenu gets unmounted and remounted every time MainFrame gets re-rendered ([here's a sandbox](https://codesandbox.io/s/rerenderdoesunmount-rx63w)). And MainFrame does get rerendered, often 2-3 times at once, with every language and project change.

I know my progress has been a bit slow recently - I've been a bit busy the past few days. I'll try to speed up now and fix this issue as soon as I can.

Reply:
First huge thanks for the sandbox :) This is a great idea to make a sandbox for these kind of issues.

You properly identified that there is an issue, which is indeed worrying because it should not happen and can create issue with listeners.

I believe the issue is because you considered renderMainMenu as a component, by renaming it to MainFrame and rendering it as a component (`<MainFrame ... />`). The problem is that the function you're passing to renderMainMenu is then considered by React as a component (a functional one). Each time you update the `foo` state, the render is called and a new function is created... in other words, it's like you are creating a new component each time this render is called! :) 

To avoid this, just don't try to make React thing it's a component. Use it as a function:
![image](https://user-images.githubusercontent.com/1280130/78820774-ad8fca00-79d8-11ea-9556-011e10db71f2.png)

No more issues:
![image](https://user-images.githubusercontent.com/1280130/78820789-b2ed1480-79d8-11ea-8cd2-1258d3182a4c.png)

This pattern is called a "Render prop". Rule of thumb: any prop called renderXXX is a render prop and should be called like a function - it's NOT a component :)